### PR TITLE
Some updates about running simulation.

### PR DESCRIPTION
**Pull request by xgzi**
## Overview
**These updates are meant to solve the following problems:**
1. Bugs about the netlist name.
2. A 'simulator.exe' was given to a simulator class, but it still alerts when initialized.
3. When running simulations in parallel, it's difficult for the user to know the correspondence between outputs and netlist files.
4. Minor revisions to convenience users.

#### 1. Bug about the netlist name.
**What happened:**

If we give a wrong netlist file name.
For example 1: `netlist=SpiceEditor("wrong_suffix.cir")`; 
Example 2: 
``` Python
runner = SimRunner()
runner.run(xxx, run_filename = "without_suffix_or_wrong_suffix")
```
There will be no warning message. (Especially, referring to this `:param run_filename: Name to be used for the log and raw file.`, it's hard for a user to know it should be with the '.net' suffix).

But in `SimRunner.run() -> self._prepare_sim(xxx) -> netlist.write_netlist(run_netlist_file)`, the netlist file name is forced to with '.net' suffix. At the same time, `run_netlist_file` remains unchanged.

So the LTspice.exe can not find "the netlist file", the simulation will not run.

**Futhermore:**
I found that the LTSpice.exe will still return 0 (succeed code) even if it cannot find the netlist file. So this program will only report that "raw and log file not found," but this is misleading.

**What I did:**
Add suffix and naming validity checking in `SimRunner._run_file_name()` to prevent the naming consistency. Corrected the documentation of param run_filename.

**Effectiveness:**
Now, it can support different netlist file suffixes but still has a warning if it's not '.net'.

#### 2. Simulator class unwanted alerts when initialized.
**What happened:**
In `ltspice_simulator.py` and `qspice_simulator.py`, if the system path cannot find corresponding executables (in Windows system paths), it will alert when it's first imported. I found it annoying.
**What I did:**
Move the alerting codes into `simulator.run()` from the class initialization.
If the user has already called `simulator.create_from()` method to specify an executable, it will not alert.
**What I did next:**
I found it inconvenient for the user to import the LTspice class. Previously the user should write like this:
```
from spicelib.simulator.ltspice_simulator import LTSpice
```
So, I optimized the convenience of direct imports.

#### 3. Difficulty in configuring parallel simulations.
**What's the problem:**
The user may use a for loop to call `SimRunner.run()` with several netlist files.
However, after all simulations are completed, the '.raw' and '.log' files may have different filenames than the user expected. 
(Because the files may be renamed to `"%s_%i%s" % ( Path(netlist).stem(), SimRunner.runno, Path(netlist).suffix()`.)
So the user cannot know the correspondence between a raw file and a netlist file.

**What I did:**
The user can use the return of `run` or `run_now` methods to build a mapping of netlist and output files.
The `SimRunner.run()` returns a RunTask instance, while the `SimRunner.run_now()` returns `(RunTask.raw, RunTask.log)`. I think the latter return is better because you said in `RunTask`: 
> """This is an internal Class and should not be used directly by the User."""

So I changed the return of `run()` to the same as `run_now()`. Now the user can know the relationship between output files and input netlists by doing like this:
``` Python
output_raw_file, output_log_file = inst_SimRunner.run(netlist = input_netlist)
in_out_mapping = {input_netlist.netlist_file: (output_raw_file, output_log_file)}
```

**There's a new problem** that although the user can know the raw filename when they run a simulation, the raw file may not be ready because the simulation may be still running in another thread or process.
So the user should wait for the simulation to complete and then do the data processing.
I updated the example file `examples/sim_runner_example.py` and added this:
``` Python
LTC = SimRunner()
...
LTC.wait_completion()
```

---
I hope these changes can improve the experience of the user.

### CODE REVIEW

File: spicelib/sim/sim_runner.py
Comment (nunobrum) [inline spicelib/sim/sim_runner.py:264]: See my comment on pull-request.

File: spicelib/sim/sim_runner.py
Comment (nunobrum) [inline spicelib/sim/sim_runner.py:397]: This is problematic. We need this on other parts of the code, such as the Client Server. 
The way it worked before, is that the SimRunner iterator would return the proper name of the raw and log files. This gave the user all he needed.

File: spicelib/sim/sim_runner.py
Comment (nunobrum) [inline spicelib/sim/sim_runner.py:337]: See Comment below

File: examples/sim_runner_example.py
Comment (nunobrum) [inline examples/sim_runner_example.py:1]: You could do what you here with the former code.
It sould suffice that you would set the run_filename parameter, and the the iterator given by the SimRunner would give back the raw and log filenames with the matching filename.

Furthermore, it had the advantage that it would return simulation results as they were completed, and you didn't have to place the wait_completion() method and only then start the processing.

Generally, I have the feeling that this does not improve the code.

File: spicelib/simulators/qspice_simulator.py
Comment (nunobrum) [inline spicelib/simulators/qspice_simulator.py:1]: I'm OK with this improvement

File: spicelib/simulators/ngspice_simulator.py
Comment (nunobrum) [inline spicelib/simulators/ngspice_simulator.py:1]: Very good! Thanks.

File: spicelib/simulators/xyce_simulator.py
Comment (nunobrum) [inline spicelib/simulators/xyce_simulator.py:1]: All good.

File: spicelib/simulators/__init__.py
Comment (nunobrum) [inline spicelib/simulators/__init__.py:1]: See comment in the pull-request level

**Review by nunobrum — CHANGES_REQUESTED** [review]
Hi @xgzi ,

Thanks a lot for your contribution. It's all clear and well explained. I agree with most of the changes, but, before I proceed with the merge, I wanted to explain in large terms my rationale. A

I've added a few comments on the code itself. But here is the summary.
**1) the netlist suffix**
Is there an imposition that the spicenetlist should be of the suffix .net. It is typically so for LTspice, but, not really so for Qspice for example. If you generate a netlist with QSpice QNX application, it will default to .cir. So, probably we shouldn't restrict it to just .net.

**2) Class initalization messages**
I am OK with the update. I was already thinking about doing something similar.

**3) simulation files naming **
I think I should have better documented this part of the code. When you used the run_filename parameter, the raw and log files used would have been used. No numbers were added.
Then you used the iterator provided by the SimRunner class which would return the same names you had provided. No need to resort to an external dictionary to do so.

** General comment **
I really appreciated the effort you placed in the changes and most of all, really liked the way you explained everything on the pull-request. I can't accept all your changes, because some of them conflict with the implementation of the simulation server.
Some of them, yes, they do make sense. 
I don't know how do you prefer to proceed. There are basically two options that I can think of:
a) I'll merge your pull-request and will cherry-pick the updates I am fine with.
b) You update your code based on my comments.
Please let me know which do you prefer, or if there is another one, please let me know.

Thanks once again and talk soon,
Nuno

File: examples/sim_runner_example.py
Comment (xgzi) [reply examples/sim_runner_example.py:1]: I agree with you.

(Just curious, is there any blocking mechanism when calling the iterator? 
I wonder if the simulation threads take too long, will the for loop drop some unfinished simulation results?

File: spicelib/__init__.py
Comment (xgzi) [reply spicelib/__init__.py:1]: I don't fully understand that.
Do you mean importing both PyLTSpice and spicelib to perform LTspice simulation while importing both Qspice and spicelib to perform Qspice simulation?
What's the advantage of it than just importing an LTspice simulator?

File: spicelib/editor/spice_editor.py
Comment (xgzi) [reply spicelib/editor/spice_editor.py:685]: I suggest changing it to:
`if self.circuit_file.suffix not in ['.net', '.cir']:`

You mentioned that different spice editor has a different suffix. I agree not to limit that.
However, I want to remind you that some pieces of the code may force a '.cir' netlist to '.net'. (In this circumstance, running simulator will fail). This might need optimization.

File: spicelib/editor/spice_editor.py
Comment (xgzi) [reply spicelib/editor/spice_editor.py:761]: I agree. I guess you want to improve the experience of building a netlist using a new set of methods. That's definitely a good idea!

However, this set of methods may need further improvement in the current stage. Some people may even feel unwilling to learn new things. 
If you allow users to build a netlist with both your method to simplify works and spice original instructions as an alternative.

**I have a compromise idea**: 
The spicelib parses instructions received by `add_instructions()`. Instead of directly appending to the netlist lines list, it identifies the ".param" keyword and then calls `set_params()` to add this command and for syntax checking.
The same idea applies to other methods like `set_component` and so on.

Comment (xgzi) [reply spicelib/editor/spice_editor.py:761]: I agree not to make a change currently.

File: spicelib/sim/run_task.py
Comment (xgzi) [reply spicelib/sim/run_task.py:111]: As I tested, if the netlist file does not exist, the `retcode` may still be zero.
This code is to make sure that the simulation indeed succeeds.

If it's `self.raw_file.exist()`, what if the previous run already generates an output, but the current run failed?

File: spicelib/simulators/__init__.py
Comment (xgzi) [reply spicelib/simulators/__init__.py:1]: I agree not to change.

File: spicelib/sim/sim_runner.py
Comment (xgzi) [reply spicelib/sim/sim_runner.py:397]: Ok.
I just wanted to let you know that you have shown another way to solve my need in the example file.
So my previous change is not necessary.

**Comment by xgzi** [comment]
Thanks for your patience and detailed comments.
All my changes are only suggestions. They can solve my problems in specific conditions, but they may negatively affect the whole project. I fully understand and appreciate you as the maintainer in taking control and deciding which part of the codes to accept.

I prefer the way that:
> "I'll merge your pull-request and will cherry-pick the updates I am fine with."
Because the problems you mentioned are valuable but need more time and effort to solve. (I refer to the suffix competitive issue and adding spice instruction issue. For the latter, I came up with an idea and explained it in my detailed comment).

As for the other problems, I think they are clear. Please see my detailed comments.

Please feel free to ask me to improve the code or discuss with me.

**Comment by nunobrum** [comment]
Hi @xgzi,

that's fine. I'll accept your pull-request and cherry pick the things that don't conflict with existing features.

Thank you,
Nuno

**Pull request merged**
